### PR TITLE
SVG primitive: Support custom aria-hidden value

### PR DESCRIPTION
**Pull request by WunderBart**
## Description

This PR introduces support for providing a custom `aria-hidden` attribute for the `SVG` primitive.

Currently, [the SVG component has the `aria-hidden` value fixed to `true`](https://github.com/WordPress/gutenberg/blob/4c472c3443513d070a50ba1e96f3a476861447b3/packages/primitives/src/svg/index.js#L29). This is fine for decorative icons that are used within a button or a link to other interactive controls. Unfortunately, this prevents us from using ARIA when we need to enhance SVG accessibility. E.g., if we wanted to add the `<title>` element inside SVG as suggested [here](https://developer.paciellogroup.com/blog/2013/12/using-aria-enhance-svg-accessibility/), it wouldn't take any effect because of the hardcoded `aria-hidden="true"` presence.

Related discussion: https://wordpress.slack.com/archives/C02RP4X03/p1585161710006700

## How has this been tested?

If explicitly specified, the `aria-hidden` prop value should be passed as is to the `aria-hidden` SVG attribute. This means that the `undefined` will also be accepted, resulting in the removal of that attribute.

⚠️ Note:
`undefined` should be the preferred way over `"false"` as it's [reported to behave inconsistently across browsers](https://www.w3.org/TR/wai-aria-1.1/#aria-hidden).

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by WunderBart** [comment]
I noticed that the [React Native implementation](https://github.com/WordPress/gutenberg/blob/89b7a6c135eff61287d8eba4975c3a542df8ac21/packages/primitives/src/svg/index.native.js) does not force the same attributes (`role`, `aria-hidden` & `focusable`). Is that intended?

File: packages/primitives/src/svg/index.js [hunk 1]
```diff
@@ -26,7 +26,9 @@ export const SVG = ( { className, isPressed, ...props } ) => {
 		className:
 			classnames( className, { 'is-pressed': isPressed } ) || undefined,
 		role: 'img',
-		'aria-hidden': 'true',
+		'aria-hidden': props.hasOwnProperty( 'aria-hidden' )
```
Comment (gziolo) [inline packages/primitives/src/svg/index.js]: Did you consider using a new prop like `isFocusable` to make it more explicit?

**Comment by aduth** [comment]
Can we just change the order of the object spread so that `...props` takes precedent? This would have the same effect, effectively allowing a developer to pass their own `aria-hidden` to override.

I believe the idea with those hard-coded values is to serve as smart defaults. But they should be just that: defaults. I don't know that we need be so opinionated that they can't be overridden.

**Comment by jeryj** [comment]
@aduth 
> Can we just change the order of the object spread so that ...props takes precedent?

I think that's a great idea 🙌 I'm not sure how best to handle the `className` one though. Should it stay after `...props` since it gets handled in a different way with the explicit `className` and `isPressed` props?

Currently:
```
const appliedProps = {
	...props,
	className: classnames( className, { 'is-pressed': isPressed } ) || undefined,
	role: 'img',
	'aria-hidden': 'true',
	focusable: 'false',
};
```

Proposed change to move `...props` lower:
```
const appliedProps = {
	 role: 'img',
	'aria-hidden': 'true',
	focusable: 'false',
	...props,
	className: classnames( className, { 'is-pressed': isPressed } ) || undefined,
};
```

**Comment by aduth** [comment]
> Proposed change to move `...props` lower:
> 
> ```
> const appliedProps = {
> 	 role: 'img',
> 	'aria-hidden': 'true',
> 	focusable: 'false',
> 	...props,
> 	className: classnames( className, { 'is-pressed': isPressed } ) || undefined,
> };
> ```

Yes, I believe your assessment here is correct. It should be last, in order to both retain the passed prop value _and_ include additional class names specific to the component.

**Comment by aduth** [comment]
Scratch that. It doesn't need to be last. Since `className` is explicitly picked, the `...props` spread value wouldn't contain it anyways, so it wouldn't override if allowed to be last.

**Comment by aduth** [comment]
Related to the ongoing Slack conversation linked in the original comment, it would be good to have demonstrated use-cases for what this is intended to be used for, not simply the fact that it would be needed if there was a hypothetical need to use the `<title>` tag.

Also note that aside from smart defaults, there is a secondary purpose of this component, which is mobile compatibility. My understanding is that usage of the default `<svg>` tag in React is not compatible in a native environment, and that this is addressed through this component in its own implementation [using a shim](https://github.com/WordPress/gutenberg/blob/8a719202ebaf6bbcbd1a18f851f54d7960d2b939/packages/primitives/src/svg/index.native.js#L11-L21).

**Comment by gziolo** [comment]
> Also note that aside from smart defaults, there is a secondary purpose of this component, which is mobile compatibility. My understanding is that usage of the default `<svg>` tag in React is not compatible in a native environment, and that this is addressed through this component in its own implementation [using a shim](https://github.com/WordPress/gutenberg/blob/8a719202ebaf6bbcbd1a18f851f54d7960d2b939/packages/primitives/src/svg/index.native.js#L11-L21).

All HTML tags and their attributes aren’t supported by React Native. See my comment: https://github.com/WordPress/gutenberg/pull/21173#discussion_r398683797. We should try to avoid platform specific language in the package containing primitives.

**Comment by aduth** [comment]
> > Also note that aside from smart defaults, there is a secondary purpose of this component, which is mobile compatibility. My understanding is that usage of the default `<svg>` tag in React is not compatible in a native environment, and that this is addressed through this component in its own implementation [using a shim](https://github.com/WordPress/gutenberg/blob/8a719202ebaf6bbcbd1a18f851f54d7960d2b939/packages/primitives/src/svg/index.native.js#L11-L21).
> 
> All HTML tags and their attributes aren’t supported by React Native. See my comment: [#21173 (comment)](https://github.com/WordPress/gutenberg/pull/21173#discussion_r398683797). We should try to avoid platform specific language in the package containing primitives.

Ah, right, I think I was making bad interpretations of the intent behind previous efforts of #9685 and #9294.

File: packages/primitives/src/svg/index.js (hunk 1 above)
Comment (WunderBart) [reply packages/primitives/src/svg/index.js]: My focus for this PR was only the `aria-hidden` attribute. Nevertheless, if we'd like to make `focusable` attribute optional with default value set, wouldn't it be better if we move it above the props spread instead of creating custom `isFocusable` prop? E.g.

```js
export const SVG = ( { className, isPressed, ...props } ) => (
	<svg
		className={
			classnames( className, { 'is-pressed': isPressed } ) || undefined
		}
		role="img"
		aria-hidden="true"
		focusable="false"
		{ ...props }
	/>
);
```

**Comment by WunderBart** [comment]
Thanks for reviewing, @gziolo @aduth @jeryj 🙌 

> Related to the ongoing Slack conversation linked in the original comment, it would be good to have demonstrated use-cases for what this is intended to be used for, not simply the fact that it would be needed if there was a hypothetical need to use the <title> tag.

By the time I created this PR we had a use-case for the `Podcast Player` block we're implementing currently in Jetpack. In the tracklist, we display a status icon for the active track:

image
Currently, we support 2 states: `playing` and `error`.

We wanted these states to be announced to user, e.g.:

> “Playing - [track name], selected”

instead of just

> “[track name], selected”

To achieve that we chose to include the `<title>` element inside, as suggested [here](https://developer.paciellogroup.com/blog/2013/12/using-aria-enhance-svg-accessibility/). It seemed to be the right way at the time, but there's a potential issue in allowing this - "[developers adding titles where not necessary](https://wordpress.slack.com/archives/C02RP4X03/p1585241528025800)". Personally, I'm not sure if we should be this opinionated here, but I'm going to trust a11y folks on this.

The alternative approach in our case was adding the visually hidden text outside the SVG, e.g.:

```js
<span class="visually-hidden">Playing: </span><svg aria-hidden="true" /> [track name]
```

...which we eventually decided to continue with (p1585251445460300-slack-ajax).

---

TL;DR, since we're not going to continue with the `title` approach and because of the mentioned concerns this way involves, I'm happy to close this PR.

cc @afercia @enriquesanchez

**Comment by afercia** [comment]
> Related to the ongoing Slack conversation linked in the original comment, it would be good to have demonstrated use-cases for what this is intended to be used for, not simply the fact that it would be needed if there was a hypothetical need to use the <title> tag.

I'd second this.

> By the time I created this PR we had a use-case for the Podcast Player block we're implementing currently in Jetpack. In the tracklist, we display a status icon for the active track:

The alternative approach with visually hidden text seems better to me, as icons are not appropriate to communicate a state in the first place. I'd agree with @WunderBart's last comment and suggest close.

Aside: re: the `focusable` attribute, worth reminding it's only used to fix an IE11 bug, where SVG icons within buttons or links become focusable themselves, thus requiring a double Tab key press to keyboard users.

File: packages/primitives/src/svg/index.js (hunk 1 above)
Comment (gziolo) [reply packages/primitives/src/svg/index.js]: Yes, it would help in this case. I'm not against this change, I just wanted to raise the awareness of the React Native and all the challenges it creates.
